pkg/handler/collector/oci/oci.go

```diff
@@ -209,63 +208,87 @@ func (o *ociCollector) fetchOCIArtifacts(ctx context.Context, repo string, rc *r
 	}
 
 	digest := manifest.GetDigest(m)
+
 	digestFormatted := fmt.Sprintf("%v-%v", digest.Algorithm(), digest.Encoded())
 	suffixList := []string{"att", "sbom"}
 	for _, suffix := range suffixList {
 		digestTag := fmt.Sprintf("%v.%v", digestFormatted, suffix)
 		// check to see if the digest + suffix has already been collected
 		if !contains(o.checkedDigest[repo], digestTag) {
 			imageTag := fmt.Sprintf("%v:%v", repo, digestTag)
-			r, err := ref.New(imageTag)
+			err = fetchOCIArtifactBlobs(ctx, rc, imageTag, docChannel)
 			if err != nil {
-				return err
+				return fmt.Errorf("failed retrieving artifact blobs from registry fallback artifacts: %w", err)
 			}
+			o.checkedDigest[repo] = append(o.checkedDigest[repo], digestTag)
+		}
+	}
 
-			// if `.att` or `.sbom`` do not exist for specified digest
-			// log error and continue
-			m, err = rc.ManifestGet(ctx, r)
-			if err != nil {
-				// this is a normal behavior, not an error when the digest does not have an attestation
-				// explicitly logging it as info to avoid call-stack when logging
-				logger.Infof("unable to get manifest for %v: %v", imageTag, err)
-				continue
-			}
+	referrerOpts := []scheme.ReferrerOpts{}
+	referrerOpts = append(referrerOpts, scheme.WithReferrerAT("application/spdx+json"))
 
-			// go through layers in reverse
-			mi, ok := m.(manifest.Imager)
-			if !ok {
-				return fmt.Errorf("reference is not a known image media type")
-			}
-			layers, err := mi.GetLayers()
-			if err != nil {
-				return err
-			}
-			for i := len(layers) - 1; i >= 0; i-- {
-				if ctx.Err() != nil {
-					return ctx.Err()
-				}
-				blob, err := rc.BlobGet(ctx, r, layers[i])
-				if err != nil {
-					return fmt.Errorf("failed pulling layer %d: %w", i, err)
-				}
-				btr1, err := blob.RawBody()
-				if err != nil {
-					return err
-				}
+	referrerList, err := rc.ReferrerList(ctx, image, referrerOpts...)
+	if err != nil {
+		return err
+	}
+	for _, referrerDesc := range referrerList.Descriptors {
+		imageTag := fmt.Sprintf("%v@%v", repo, referrerDesc.Digest.String())
```

Comment:
Looks like we need a check similar to:
```
if !contains(o.checkedDigest[repo], digestTag) {
			imageTag := fmt.Sprintf("%v:%v", repo, digestTag)
			err = fetchOCIArtifactBlobs(ctx, rc, imageTag, docChannel)
			if err != nil {
				return fmt.Errorf("failed retrieving artifact blobs from registry fallback artifacts: %w", err)
			}
			o.checkedDigest[repo] = append(o.checkedDigest[repo], digestTag)
		}
```
so that we do not keep rechecking if already captured?

Reply:
To answer your question: "Check if idempotence matters - will an artifact that's collected twice be duplicated?"

- if collected twice and the `blob` that is collected is the same, it will not be duplicated. For example, if an SBOM is collected, it will parse into the same `hasSBOM`, `packages`, `artifacts`, `dependency`, and `occurrence` nodes.